File: Credicrew/backend/app/services/match.py
"""Explainable match engine (Python mirror of frontend/src/lib/match.ts).

Given a free-text job query, score each candidate on skill coverage,
location, and seniority, returning a per-factor breakdown so the API
consumer can explain *why* the candidate matched.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field, asdict
from typing import Any, Iterable

ALIASES: dict[str, str] = {
    "reactjs": "react", "react.js": "react",
    "nextjs": "next.js",
    "nodejs": "node.js", "node": "node.js",
    "ts": "typescript",
    "js": "javascript",
    "py": "python",
    "postgresql": "postgres", "psql": "postgres",
    "fast api": "fastapi",
    "tailwindcss": "tailwind",
    "mongo": "mongodb",
    "k8s": "kubernetes",
    "golang": "go",
    "bangalore": "bengaluru",
}

SKILL_VOCAB: set[str] = {
    "react", "next.js", "vue", "svelte", "angular",
    "typescript", "javascript", "python", "go", "rust", "java", "kotlin", "swift",
    "fastapi", "flask", "django", "express", "nest.js", "spring",
    "postgres", "mysql", "mongodb", "redis", "kafka", "rabbitmq",
    "aws", "gcp", "azure", "docker", "kubernetes", "terraform",
    "graphql", "rest", "grpc",
    "tailwind", "css", "html", "sass",
    "node.js",
    "pandas", "numpy", "pytorch", "tensorflow", "scikit-learn", "ml", "nlp", "llm",
    "prisma", "sqlalchemy",
    "cypress", "jest", "playwright", "pytest",
}

LOCATION_VOCAB: set[str] = {
    "bengaluru", "mumbai", "delhi", "hyderabad", "chennai",
    "pune", "kolkata", "noida", "gurgaon", "ahmedabad", "kochi",
    "remote", "hybrid", "onsite",
}

SENIORITY = ("intern", "junior", "mid", "senior", "staff", "principal", "lead")

WEIGHTS = {"skill": 0.55, "loc": 0.15, "sen": 0.20, "base": 0.10}


@dataclass
class QueryPlan:
    text: str
    skills: list[str]
    location: str | None = None
    seniority: str | None = None


@dataclass
class MatchFactor:
    key: str
    label: str
    impact: int


@dataclass
class MatchResult:
    score: int
    matched_skills: list[str] = field(default_factory=list)
    missing_skills: list[str] = field(default_factory=list)
    seniority_wanted: str | None = None
    seniority_candidate: str | None = None
    seniority_match: bool = True
    location_wanted: str | None = None
    location_match: str = "full"
    factors: list[MatchFactor] = field(default_factory=list)


def _canon(token: str) -> str:
    t = re.sub(r"^[\W_]+|[\W_]+$", "", token.lower())
    return ALIASES.get(t, t)


def _tokens(text: str) -> list[str]:
    return [
        _canon(t)
        for t in re.split(r"\s+", re.sub(r"[()\[\]{},;/]", " ", text.lower()))
        if t
    ]


def extract_skills(text: str) -> list[str]:
    if not text:
        return []
    out: set[str] = set()
    for t in _tokens(text):
        if t in SKILL_VOCAB:
            out.add(t)
    lower = text.lower()
    for skill in SKILL_VOCAB:
        if ("." in skill or " " in skill) and skill in lower:
            out.add(skill)
    for phrase, target in ALIASES.items():
        if " " in phrase and phrase in lower and target in SKILL_VOCAB:
            out.add(target)
    return sorted(out)


def extract_location(text: str) -> str | None:
    if not text:
        return None
    m = re.search(r"\bin\s+([A-Za-z][A-Za-z\s]+?)(?:$|[.,;])", text, re.IGNORECASE)
    if m:
        loc = _canon(m.group(1).strip())
        if loc in LOCATION_VOCAB:
            return loc
    for t in _tokens(text):
        if t in LOCATION_VOCAB:
            return t
    return None


def extract_seniority(text: str) -> str | None:
    if not text:
        return None
    lower = text.lower()
    for s in SENIORITY:
        if re.search(rf"\b{s}\b", lower):
            return s
    return None


def plan_query(text: str) -> QueryPlan:
    return QueryPlan(
        text=text,
        skills=extract_skills(text),
        location=extract_location(text),
        seniority=extract_seniority(text),
    )


def match_candidate(plan: QueryPlan, c: dict[str, Any]) -> MatchResult:
    bag: set[str] = set()
    for src in (c.get("tags") or [], c.get("keywords") or [], (c.get("role") or "").split()):
        for t in src:
            ct = _canon(t)
            if ct:
                bag.add(ct)

    cand_loc_raw = (c.get("location") or "").split("(")[0]
    cand_loc = _canon(cand_loc_raw.strip())

    matched: list[str] = []
    missing: list[str] = []
    for s in plan.skills:
        (matched if s in bag else missing).append(s)
    skill_cov = 0.8 if not plan.skills else len(matched) / len(plan.skills)

    if not plan.location:
        loc_state = "full"
    elif cand_loc == plan.location or plan.location == "remote":
        loc_state = "full"
    elif cand_loc == "remote" or "hybrid" in cand_loc:
        loc_state = "partial"
    else:
        loc_state = "none"
    loc_score = {"full": 1.0, "partial": 0.5, "none": 0.0}[loc_state]

    cand_sen = extract_seniority(" ".join([c.get("role") or "", c.get("headline") or ""]))
    sen_match = plan.seniority == cand_sen if plan.seniority else True
    sen_score = (
        1.0 if not plan.seniority
        else 1.0 if sen_match
        else 0.3 if cand_sen
        else 0.6
    )

    raw = (
        WEIGHTS["skill"] * skill_cov
        + WEIGHTS["loc"] * loc_score
        + WEIGHTS["sen"] * sen_score
        + WEIGHTS["base"]
    )
    score = round(max(0.0, min(1.0, raw)) * 100)

    factors: list[MatchFactor] = [
        MatchFactor("base", "Baseline fit", round(WEIGHTS["base"] * 100)),
    ]
    if plan.skills:
        factors.append(MatchFactor(
            "skills",
            f"{len(matched)}/{len(plan.skills)} required skill{'' if len(plan.skills)==1 else 's'}",
            round(WEIGHTS["skill"] * 100 * skill_cov),
        ))
    else:
        factors.append(MatchFactor(
            "skills", "No specific skills requested",
            round(WEIGHTS["skill"] * 100 * skill_cov),
        ))
    if plan.location:
        label = (
            f"Location · {plan.location}" if loc_state == "full"
            else f"Location · {c.get('location') or cand_loc} (flex)" if loc_state == "partial"
            else f"Location · {c.get('location') or 'unknown'} (mismatch)"
        )
        factors.append(MatchFactor("location", label, round(WEIGHTS["loc"] * 100 * loc_score)))
    if plan.seniority:
        label = (
            f"Seniority · {plan.seniority}" if sen_match
            else f"Seniority · {cand_sen or 'unspecified'} (wanted {plan.seniority})"
        )
        factors.append(MatchFactor("seniority", label, round(WEIGHTS["sen"] * 100 * sen_score)))

    return MatchResult(
        score=score,
        matched_skills=matched,
        missing_skills=missing,
        seniority_wanted=plan.seniority,
        seniority_candidate=cand_sen,
        seniority_match=sen_match,
        location_wanted=plan.location,
        location_match=loc_state,
        factors=factors,
    )


def rank(plan: QueryPlan, candidates: Iterable[dict[str, Any]]) -> list[dict[str, Any]]:
    results = []
    for c in candidates:
        r = match_candidate(plan, c)
        results.append({
            "candidate_id": c.get("id"),
            "name": c.get("name"),
            "match": {
                "score": r.score,
                "matched_skills": r.matched_skills,
                "missing_skills": r.missing_skills,
                "seniority": {
                    "wanted": r.seniority_wanted,
                    "candidate": r.seniority_candidate,
                    "match": r.seniority_match,
                },
                "location": {"wanted": r.location_wanted, "match": r.location_match},
                "factors": [asdict(f) for f in r.factors],
            },
        })
    results.sort(key=lambda x: -x["match"]["score"])
    return results

File: Credicrew/frontend/src/lib/match.ts
// Explainable match engine.
//
// Takes a free-text job query like
//   "Senior backend (FastAPI + Postgres) in Bengaluru"
// and scores a candidate against it with a breakdown of why.
//
// Pure function, no network. Same scoring logic is mirrored on the backend
// in `app/services/match.py` so server-side results agree.

export type MatchFactor = {
  key: "skills" | "location" | "seniority" | "base";
  label: string;
  impact: number; // points this factor contributed (0..100)
};

export type MatchResult = {
  score: number; // 0..100
  matchedSkills: string[];
  missingSkills: string[];
  seniority: { wanted?: string; candidate?: string; match: boolean };
  location: { wanted?: string; match: "full" | "partial" | "none" };
  factors: MatchFactor[];
};

export type CandidateLike = {
  name?: string;
  role?: string;
  location?: string;
  tags?: string[];
  keywords?: string[];
  headline?: string;
};

export type QueryPlan = {
  text: string;
  skills: string[];
  location?: string;
  seniority?: string;
};

const ALIASES: Record<string, string> = {
  reactjs: "react",
  "react.js": "react",
  nextjs: "next.js",
  nodejs: "node.js",
  node: "node.js",
  ts: "typescript",
  js: "javascript",
  py: "python",
  postgresql: "postgres",
  psql: "postgres",
  "fast api": "fastapi",
  tailwindcss: "tailwind",
  mongo: "mongodb",
  k8s: "kubernetes",
  golang: "go",
  bangalore: "bengaluru",
};

const SKILL_VOCAB = new Set<string>([
  "react", "next.js", "vue", "svelte", "angular",
  "typescript", "javascript", "python", "go", "rust", "java", "kotlin", "swift",
  "fastapi", "flask", "django", "express", "nest.js", "spring",
  "postgres", "mysql", "mongodb", "redis", "kafka", "rabbitmq",
  "aws", "gcp", "azure", "docker", "kubernetes", "terraform",
  "graphql", "rest", "grpc",
  "tailwind", "css", "html", "sass",
  "node.js",
  "pandas", "numpy", "pytorch", "tensorflow", "scikit-learn", "ml", "nlp", "llm",
  "prisma", "sqlalchemy",
  "cypress", "jest", "playwright", "pytest",
]);

const LOCATION_VOCAB = new Set<string>([
  "bengaluru", "mumbai", "delhi", "hyderabad", "chennai",
  "pune", "kolkata", "noida", "gurgaon", "ahmedabad", "kochi",
  "remote", "hybrid", "onsite",
]);

const SENIORITY = ["intern", "junior", "mid", "senior", "staff", "principal", "lead"];

const WEIGHTS = { skill: 0.55, loc: 0.15, sen: 0.2, base: 0.1 };

function canon(token: string): string {
  const t = token.toLowerCase().replace(/^[\W_]+|[\W_]+$/g, "");
  return ALIASES[t] ?? t;
}

function tokens(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[()[\]{},;/]/g, " ")
    .split(/\s+/)
    .map(canon)
    .filter(Boolean);
}

export function extractSkills(text: string): string[] {
  if (!text) return [];
  const out = new Set<string>();
  for (const t of tokens(text)) if (SKILL_VOCAB.has(t)) out.add(t);
  // multi-word phrase lookup
  const lower = text.toLowerCase();
  for (const skill of SKILL_VOCAB) {
    if (skill.includes(".") || skill.includes(" ")) {
      if (lower.includes(skill)) out.add(skill);
    }
  }
  // alias phrase lookup
  for (const phrase of Object.keys(ALIASES)) {
    if (phrase.includes(" ") && lower.includes(phrase)) {
      const target = ALIASES[phrase];
      if (SKILL_VOCAB.has(target)) out.add(target);
    }
  }
  return Array.from(out);
}

export function extractLocation(text: string): string | undefined {
  if (!text) return undefined;
  const m = text.match(/\bin\s+([A-Za-z][A-Za-z\s]+?)(?:$|[.,;])/i);
  if (m) {
    const loc = canon(m[1].trim());
    if (LOCATION_VOCAB.has(loc)) return loc;
  }
  for (const t of tokens(text)) if (LOCATION_VOCAB.has(t)) return t;
  return undefined;
}

export function extractSeniority(text: string): string | undefined {
  if (!text) return undefined;
  const lower = text.toLowerCase();
  for (const s of SENIORITY) if (new RegExp(`\\b${s}\\b`).test(lower)) return s;
  return undefined;
}

export function planQuery(text: string): QueryPlan {
  return {
    text,
    skills: extractSkills(text),
    location: extractLocation(text),
    seniority: extractSeniority(text),
  };
}

export function matchCandidate(plan: QueryPlan, c: CandidateLike): MatchResult {
  const bag = new Set<string>(
    [
      ...(c.tags ?? []),
      ...(c.keywords ?? []),
      ...((c.role ?? "").split(/\s+/)),
    ]
      .map(canon)
      .filter(Boolean)
  );
  const candLoc = canon((c.location ?? "").split(/[(),]/)[0]);

  const matched: string[] = [];
  const missing: string[] = [];
  for (const s of plan.skills) {
    if (bag.has(s)) matched.push(s);
    else missing.push(s);
  }
  const skillCov =
    plan.skills.length === 0 ? 0.8 : matched.length / plan.skills.length;

  let locState: "full" | "partial" | "none";
  if (!plan.location) locState = "full";
  else if (candLoc === plan.location || plan.location === "remote")
    locState = "full";
  else if (candLoc === "remote" || candLoc.includes("hybrid"))
    locState = "partial";
  else locState = "none";
  const locScore = locState === "full" ? 1 : locState === "partial" ? 0.5 : 0;

  const candSen = extractSeniority([c.role ?? "", c.headline ?? ""].join(" "));
  const senMatch = plan.seniority ? candSen === plan.seniority : true;
  const senScore = !plan.seniority
    ? 1
    : senMatch
    ? 1
    : candSen
    ? 0.3
    : 0.6;

  const raw =
    WEIGHTS.skill * skillCov +
    WEIGHTS.loc * locScore +
    WEIGHTS.sen * senScore +
    WEIGHTS.base;
  const score = Math.round(Math.max(0, Math.min(1, raw)) * 100);

  const factors: MatchFactor[] = [];
  factors.push({
    key: "base",
    label: "Baseline fit",
    impact: Math.round(WEIGHTS.base * 100),
  });
  if (plan.skills.length) {
    factors.push({
      key: "skills",
      label: `${matched.length}/${plan.skills.length} required skill${
        plan.skills.length === 1 ? "" : "s"
      }`,
      impact: Math.round(WEIGHTS.skill * 100 * skillCov),
    });
  } else {
    factors.push({
      key: "skills",
      label: "No specific skills requested",
      impact: Math.round(WEIGHTS.skill * 100 * skillCov),
    });
  }
  if (plan.location) {
    const label =
      locState === "full"
        ? `Location · ${plan.location}`
        : locState === "partial"
        ? `Location · ${c.location ?? candLoc} (flex)`
        : `Location · ${c.location ?? "unknown"} (mismatch)`;
    factors.push({
      key: "location",
      label,
      impact: Math.round(WEIGHTS.loc * 100 * locScore),
    });
  }
  if (plan.seniority) {
    factors.push({
      key: "seniority",
      label: senMatch
        ? `Seniority · ${plan.seniority}`
        : `Seniority · ${candSen ?? "unspecified"} (wanted ${plan.seniority})`,
      impact: Math.round(WEIGHTS.sen * 100 * senScore),
    });
  }

  return {
    score,
    matchedSkills: matched,
    missingSkills: missing,
    seniority: { wanted: plan.seniority, candidate: candSen, match: senMatch },
    location: { wanted: plan.location, match: locState },
    factors,
  };
}

export function scoreBand(score: number): "strong" | "solid" | "weak" {
  if (score >= 80) return "strong";
  if (score >= 60) return "solid";
  return "weak";
}
